File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
import java.util.*;
import java.math.*;

class Solution {
    
    public boolean canEat(int[] piles, int h, int speed) {
        BigInteger time = BigInteger.valueOf(0);
        for(int pile : piles) {
            if(pile <= speed) {
                time = time.add(BigInteger.valueOf(1));
            } else {
                time = time.add(BigInteger.valueOf(pile / speed));
                if(pile % speed != 0) {
                    time = time.add(BigInteger.valueOf(1));
                }
            }
        }
        //System.out.println(time);
        //System.out.println(time.compareTo(BigInteger.valueOf(h)));
        return time.compareTo(BigInteger.valueOf(h)) <= 0;
    }
    
    public int minEatingSpeed(int[] piles, int h) {
        
        int low = 1;
        int high = 1;
        
        for(int pile: piles) {
            high = high < pile ? pile : high;
        }
        
        int ans = high;
        while(low <= high) {
            int speed = low + (high - low) / 2;
            if(canEat(piles, h, speed)) {
                ans = ans > speed ? speed: ans;
                high = speed - 1;
            } else {
                low = speed + 1;
            }
        }
        return ans;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    
    let min_speed = 1, max_speed = Math.max(...piles);
    
    let canEat = (speed) => {
        let time = 0
        for(let pile of piles) {
            if(pile <= speed) {
                time++
            } else {
                time = time + parseInt(pile / speed);
                if(pile % speed != 0) {
                    time++
                }
            }
        }
        return time <= h;
    }
    
    let ans = max_speed;
    let low = min_speed, high = max_speed;
    while(low <= high) {
        let mid = low + parseInt((high - low) / 2);
        if(canEat(mid)) {
            ans = ans > mid ? mid : ans;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return ans;
};